File: webui/src/main/java/com/taskadapter/webui/pages/ConfigsPage.java
package com.taskadapter.webui.pages;

import com.taskadapter.web.uiapi.UISyncConfig;
import com.taskadapter.webui.ConfigOperations;
import com.taskadapter.webui.Page;
import com.taskadapter.webui.Sizes;
import com.taskadapter.webui.Tracker;
import com.vaadin.event.FieldEvents;
import com.vaadin.ui.Alignment;
import com.vaadin.ui.Button;
import com.vaadin.ui.HorizontalLayout;
import com.vaadin.ui.Html5File;
import com.vaadin.ui.Label;
import com.vaadin.ui.TextField;
import com.vaadin.ui.VerticalLayout;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

import static com.vaadin.server.Sizeable.Unit.PIXELS;

public final class ConfigsPage {

    /**
     * Callback for config list page.
     */
    public interface Callback {
        /**
         * User requested to edit config.
         *
         * @param config selected config.
         */
        void edit(UISyncConfig config);

        /**
         * User requested synchronization in "forward" directions (from left to
         * right).
         *
         * @param config config for the operation.
         */
        void forwardSync(UISyncConfig config);

        /**
         * User requested synchronization in "reverse" direction (from right to
         * left).
         *
         * @param config config for the operation.
         */
        void backwardSync(UISyncConfig config);

        /**
         * Performs a forward drop-in.
         *
         * @param config config to use.
         * @param file   file to receive.
         */
        void forwardDropIn(UISyncConfig config, Html5File file);

        /**
         * Performs a backward drop-in.
         *
         * @param config config to use.
         * @param file   file to receive.
         */
        void backwardDropIn(UISyncConfig config, Html5File file);

        /**
         * User requested creation of a new config.
         */
        void newConfig();
    }

    /**
     * Config display mode.
     */
    public enum DisplayMode {
        /**
         * This page is displaying only owned configs.
         */
        OWNED_CONFIGS {
            String nameOf(UISyncConfig config) {
                return config.getLabel();
            }
        },

        /**
         * This page is displaying all configs.
         */
        ALL_CONFIGS {
            String nameOf(UISyncConfig config) {
                return config.getOwnerName() + " : " + config.getLabel();
            }
        };

        /**
         * Provides a name of the config.
         */
        abstract String nameOf(UISyncConfig config);
    }

    /**
     * Comparator for configuration files.
     */
    private static final class ConfigComparator implements
            Comparator<UISyncConfig> {
        /**
         * Name of the current user.
         */
        private final String userName;

        ConfigComparator(String userName) {
            this.userName = userName;
        }

        @Override
        public int compare(UISyncConfig o1, UISyncConfig o2) {
            final boolean isMyConfig1 = userName.equals(o1.getOwnerName());
            final boolean isMyConfig2 = userName.equals(o2.getOwnerName());

            if (isMyConfig1 != isMyConfig2)
                return isMyConfig1 ? -1 : 1;

            final int ucomp = o1.getOwnerName().compareTo(o2.getOwnerName());
            if (ucomp != 0)
                return ucomp;

            return o1.getLabel().compareTo(o2.getLabel());
        }

    }

    public VerticalLayout layout;
    private Tracker tracker;
    private Collection<UISyncConfig> configs;
    private Boolean showAll;
    private final DisplayMode displayMode;
    private final Callback callback;
    private ConfigOperations configOperations;
    private Runnable showSavedResults;
    private final VerticalLayout configsLayout;
    TextField filterField = new TextField();

    public ConfigsPage(Tracker tracker, Boolean showAll,
                       final Callback callback, ConfigOperations configOperations,
                Runnable showSavedResults) {

        this.tracker = tracker;
        this.showAll = showAll;
        this.displayMode = showAll ? ConfigsPage.DisplayMode.ALL_CONFIGS
                : ConfigsPage.DisplayMode.OWNED_CONFIGS;
        this.callback = callback;
        this.configOperations = configOperations;
        this.showSavedResults = showSavedResults;

        layout = new VerticalLayout();
        layout.setSpacing(true);
        layout.setWidth(560, PIXELS);

        final HorizontalLayout actionPanel = new HorizontalLayout();
        actionPanel.setWidth(Sizes.configsListWidth());
        layout.addComponent(actionPanel);

        final Button addButton = new Button(Page.message("configsPage.buttonNewConfig"));
        addButton.addClickListener((Button.ClickListener) clickEvent -> callback.newConfig());
        actionPanel.addComponent(addButton);
        actionPanel.setComponentAlignment(addButton, Alignment.MIDDLE_LEFT);

        final HorizontalLayout filterPanel = new HorizontalLayout();
        filterField.addTextChangeListener((FieldEvents.TextChangeListener) event -> filterFields(event.getText()));
        filterPanel.addComponent(new Label(Page.message("configsPage.filter")));
        filterPanel.addComponent(filterField);
        filterPanel.setSpacing(true);
        actionPanel.addComponent(filterPanel);
        actionPanel.setComponentAlignment(filterPanel, Alignment.MIDDLE_RIGHT);

        configsLayout = new VerticalLayout();
        configsLayout.setSpacing(true);
        configsLayout.setWidth(560, PIXELS);
        refreshConfigs();
        layout.addComponent(configsLayout);
    }

    private void refreshConfigs() {
        final List<UISyncConfig> configs = showAll ? configOperations.getManageableConfigs() : configOperations.getOwnedConfigs();
        final List<UISyncConfig> configsCopy = new ArrayList<>(configs);
        Collections.sort(configsCopy, new ConfigComparator(configOperations.userName()));
        this.configs = configsCopy;
        setDisplayedConfigs(configs);
        filterFields(filterField.getValue());
    }

    private void setDisplayedConfigs(Collection<UISyncConfig> dispConfigs) {
        configsLayout.removeAllComponents();

        for (UISyncConfig config : dispConfigs) {
            configsLayout.addComponent(ConfigActionsPanel.render(config,
                    displayMode, callback, configOperations, this::refreshConfigs,
                    showSavedResults,
                    tracker)
            );
        }
    }

    private void filterFields(String filterStr) {
        final String[] words = filterStr == null ? new String[0] : filterStr
                .toLowerCase().split(" +");
        final List<UISyncConfig> res = configs.stream()
                .filter(config -> matches(config, words))
                .collect(Collectors.toList());

        setDisplayedConfigs(res);
    }

    private boolean matches(UISyncConfig config, String[] filters) {
        if (filters.length == 0) {
            return true;
        }
        for (String name : filters) {
            final String confName = displayMode.nameOf(config);
            if (!confName.toLowerCase().contains(name)
                    && !config.getConnector1().getLabel().toLowerCase()
                    .contains(name)
                    && !config.getConnector2().getLabel().toLowerCase()
                    .contains(name)) {
                return false;
            }
        }
        return true;
    }
}

File: webui/src/main/java/com/taskadapter/webui/pages/ConfigsPage.scala
package com.taskadapter.webui.pages

import java.util.Comparator

import com.taskadapter.web.uiapi.{ConfigId, UISyncConfig}
import com.taskadapter.webui.{ConfigOperations, Page, Tracker}
import com.vaadin.ui._

object ConfigsPage {

  /**
    * Callback for config list page.
    */
  trait Callback {
    /**
      * User requested to edit config.
      *
      * @param config selected config.
      */
    def edit(config: UISyncConfig): Unit

    /**
      * User requested synchronization in "forward" directions (from left to
      * right).
      *
      * @param config config for the operation.
      */
    def forwardSync(config: UISyncConfig): Unit

    /**
      * User requested synchronization in "reverse" direction (from right to
      * left).
      *
      * @param config config for the operation.
      */
    def backwardSync(config: UISyncConfig): Unit

    /**
      * Performs a forward drop-in.
      *
      * @param config config to use.
      * @param file   file to receive.
      */
    def forwardDropIn(config: UISyncConfig, file: Html5File): Unit

    /**
      * Performs a backward drop-in.
      *
      * @param config config to use.
      * @param file   file to receive.
      */
    def backwardDropIn(config: UISyncConfig, file: Html5File): Unit

    /**
      * User requested creation of a new config.
      */
    def newConfig(): Unit

    def showAllPreviousResults(configId: ConfigId): Unit

    def showLastExportResult(configId: ConfigId): Unit
  }

  /**
    * Comparator for configuration files.
    */
  class ConfigComparator(
                          /**
                            * Name of the current user.
                            */
                          val userName: String) extends Comparator[UISyncConfig] {
    override def compare(o1: UISyncConfig, o2: UISyncConfig): Int = {
      val isMyConfig1 = userName == o1.getOwnerName
      val isMyConfig2 = userName == o2.getOwnerName
      if (isMyConfig1 != isMyConfig2) return if (isMyConfig1) -1
      else 1
      val ucomp = o1.getOwnerName.compareTo(o2.getOwnerName)
      if (ucomp != 0) return ucomp
      o1.getLabel.compareTo(o2.getLabel)
    }
  }

}

class ConfigsPage(tracker: Tracker, showAll: Boolean, callback: ConfigsPage.Callback, configOperations: ConfigOperations) {
  val displayMode = if (showAll) DisplayMode.ALL_CONFIGS
  else DisplayMode.OWNED_CONFIGS

  private var configs = Seq[UISyncConfig]()

  val layout = new VerticalLayout
  layout.setSpacing(true)
  val actionPanel = new HorizontalLayout
  actionPanel.setWidth("100%")
  val addButton = new Button(Page.message("configsPage.buttonNewConfig"))
  addButton.addClickListener(_ => callback.newConfig())
  actionPanel.addComponent(addButton)
  actionPanel.setComponentAlignment(addButton, Alignment.MIDDLE_LEFT)

  val filterPanel = new HorizontalLayout
  val filterField = new TextField
  filterField.addTextChangeListener(e => filterFields(e.getText))
  filterPanel.addComponent(new Label(Page.message("configsPage.filter")))
  filterPanel.addComponent(filterField)
  filterPanel.setSpacing(true)
  actionPanel.addComponent(filterPanel)
  actionPanel.setComponentAlignment(filterPanel, Alignment.MIDDLE_RIGHT)
  val configsTopLevelLayout = new HorizontalLayout()
  configsTopLevelLayout.setSpacing(true)
  configsTopLevelLayout.setSizeFull()

  val configsLayout = new HorizontalLayout()
  configsLayout.setWidth("100%")

  val listSelect = configureListSelect()
  val listPanel = new Panel(listSelect)
  listPanel.setWidth("300px")
  configsLayout.addComponent(listPanel)

  val configArea = new Panel()
  configsLayout.addComponent(configArea)
  configsLayout.setExpandRatio(configArea, 1.0f)

  configsTopLevelLayout.addComponent(configsLayout)
  configsTopLevelLayout.setComponentAlignment(configsLayout, Alignment.TOP_CENTER)

  layout.addComponent(actionPanel)
  layout.addComponent(configsTopLevelLayout)
  layout.setComponentAlignment(actionPanel, Alignment.TOP_LEFT)
  refreshConfigs()

  private def refreshConfigs() = {
    val loadedConfigs = if (showAll) configOperations.getManageableConfigs
    else configOperations.getOwnedConfigs
    configs = loadedConfigs.sortBy(c => c.getOwnerName)
    setDisplayedConfigs(configs)
    filterFields(filterField.getValue)
  }

  private def setDisplayedConfigs(dispConfigs: Seq[UISyncConfig]): Unit = {
    dispConfigs.foreach(config => {
      listSelect.addItem(config.id)
      listSelect.setItemCaption(config.id, config.label)
    })
  }

  def showConfigInfo(configId: ConfigId): Unit = {
    val maybeConfig = configOperations.getConfig(configId)
    if (maybeConfig.isDefined) {
      val component = ConfigActionsPanel.render(maybeConfig.get, displayMode, callback, configOperations,
        () => refreshConfigs,
        () => callback.showAllPreviousResults(configId),
        () => callback.showLastExportResult(configId), tracker)
      component.setMargin(true)
      configArea.setContent(component)
    }
  }

  private def configureListSelect(): ListSelect = {
    val listSelect = new ListSelect()
    listSelect.setNullSelectionAllowed(false)
    listSelect.setWidth("100%")
    listSelect.setImmediate(true)
    listSelect.addValueChangeListener(e => {
      showConfigInfo(e.getProperty.getValue.asInstanceOf[ConfigId])
    })
    listSelect
  }

  private def filterFields(filterStr: String): Unit = {
    val words = if (filterStr == null) new Array[String](0)
    else filterStr.toLowerCase.split(" +")
    val res = configs.filter(config => matches(config, words))
    setDisplayedConfigs(res)
  }

  private def matches(config: UISyncConfig, filters: Array[String]): Boolean = {
    if (filters.length == 0) return true
    for (name <- filters) {
      val confName = displayMode.nameOf(config)
      if (!confName.toLowerCase.contains(name) && !config.getConnector1.getLabel.toLowerCase.contains(name) && !config.getConnector2.getLabel.toLowerCase.contains(name)) return false
    }
    true
  }

  def ui = layout
}